libraries/access.rb

```diff
@@ -0,0 +1,293 @@
+#
+# Cookbook:: postgresql
+# Library:: access
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+module PostgreSQL
+  module Cookbook
+    module AccessHelpers
+      module PgHbaTemplate
+        def config_resource_init
+          config_resource_create unless config_resource_exist?
+        end
+
+        def config_resource
+          return unless config_resource_exist?
+
+          find_resource!(:template, new_resource.config_file)
+        end
+
+        private
+
+        def config_resource_exist?
+          !find_resource!(:template, new_resource.config_file).nil?
+        rescue Chef::Exceptions::ResourceNotFound
+          false
+        end
+
+        def config_resource_create
+          file = PostgreSQL::Cookbook::AccessHelpers::PgHba::PgHbaFile.new
+          file.read!(new_resource.config_file) if ::File.exist?(new_resource.config_file)
+
+          with_run_context(:root) do
+            declare_resource(:template, new_resource.config_file) do
+              cookbook new_resource.cookbook
+              source new_resource.source
+
+              owner new_resource.owner
+              group new_resource.group
+              mode new_resource.filemode
+              sensitive new_resource.sensitive
+
+              variables(pg_hba: file)
+
+              action :nothing
+              delayed_action :create
+            end
+          end
+        end
+      end
+
+      module PgHba
+        ENTRY_TYPES = %w(local host hostssl hostnossl hostgssenc hostnogssenc).freeze
+
+        class PgHbaFile
+          attr_reader :entries
+
+          def initialize
+            @entries = []
+            @access_entries = []
+          end
+
+          def read!(file = 'pg_hba.conf', sort: true)
```

Comment:
Is read! Going to modify the file?

Reply:
Other way around it'll load from the file and clobber whatever was in the object before

Reply:
If it isn't going to modify it then it doesn't need the !

Reply:
Hmm i'm not so sure as it modifies the receiver by replacing any entries in the object with the ones in the file, removing any changes. The same as `:map!` will change whatever it's called upon.